src/main/kotlin/io/cloudflight/gradle/autoconfigure/AutoConfigureGradlePlugin.kt

```diff
@@ -0,0 +1,52 @@
+package io.cloudflight.gradle.autoconfigure
+
+import io.cloudflight.gradle.autoconfigure.extentions.gradle.api.plugins.apply
+import io.cloudflight.gradle.autoconfigure.extentions.gradle.api.plugins.create
+import io.cloudflight.gradle.autoconfigure.extentions.gradle.api.plugins.getByType
+import io.cloudflight.gradle.autoconfigure.java.*
+import io.cloudflight.gradle.autoconfigure.java.isJavaProject
+import io.cloudflight.gradle.autoconfigure.util.isServerProject
+import org.gradle.api.GradleException
+import org.gradle.api.JavaVersion
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.provider.Provider
+
+class AutoConfigureGradlePlugin : Plugin<Project> {
+    override fun apply(target: Project) {
+        if (target != target.rootProject) throw GradleException("'autoconfigure-gradle' plugin can only be applied to the root project.")
+
+        val autoConfigure = target.extensions.create(EXTENSION_NAME, AutoConfigureExtension::class)
+        with(autoConfigure.java) {
+            javaVersion.convention(JAVA_VERSION)
+            encoding.convention(JAVA_ENCODING)
+            vendorName.convention(VENDOR_NAME)
+            serverProjectSuffix.convention("-server")
+        }
+
+        for (project in target.allprojects) {
+            applyPlugins(project, autoConfigure)
+        }
+
+    }
+
+    companion object {
+        const val EXTENSION_NAME = "autoConfigure"
+    }
+
+    private fun applyPlugins(project: Project, autoConfigure: AutoConfigureExtension) {
+        val plugins = project.plugins
+
+        if (isJavaProject(project)) {
+            plugins.apply(JavaConfigurePlugin::class)
+            val extension = project.extensions.getByType(JavaConfigurePluginExtension::class)
+            val javaConfigure = autoConfigure.java
+            extension.apply {
+                javaVersion.set(javaConfigure.javaVersion)
```

Comment:
The user config is winning as you can see in the test-fixture `autoconfigure/single-java-module-configure` because all code of applying a plugin is run before the execution of the gradle script is continued, which will then override these values again.

But if you can think of any scenario that might not work, then I can also override the `convention` here.    
While implementing I just thought that its weird to override the convention of another plugin.